File: sliding_window/k_distinct_subarrays/k_distinct_subarrays.py

def _count_limited_subarrays(nums: list[int], distinction_limit: int) -> int:
    total_subarrays = 0

    nums2counts, start_idx = dict(), 0
    for end_idx, num in enumerate(nums):
        if num not in nums2counts.keys():
            nums2counts.update({num: 0})
        nums2counts[num] += 1

        while len(nums2counts) > distinction_limit:
            nums2counts[nums[start_idx]] -= 1
            if nums2counts[nums[start_idx]] == 0:
                del nums2counts[nums[start_idx]]
            start_idx += 1

        total_subarrays += end_idx - start_idx + 1

    return total_subarrays

def count_k_distinct_subarrays(nums: list[int], k: int) -> int:  # LeetCode Q.992.
    return _count_limited_subarrays(nums, k) - _count_limited_subarrays(nums, k - 1)

File: sliding_window/0992_hard/k_distinct_subarrays.cpp
#include <vector>
#include <unordered_map>
using namespace std;

int count_limited_subarrays(vector<int> &nums, int distinction_limit)
{
    int total_subarrays = 0;

    unordered_map<int, int> nums2counts;
    int left_idx = 0;
    for (int right_idx = 0; right_idx < nums.size(); right_idx++)
    {
        nums2counts[nums[right_idx]]++;
        while (nums2counts.size() > distinction_limit)
        {
            nums2counts[nums[left_idx]]--;
            if (nums2counts[nums[left_idx]] == 0)
            {
                nums2counts.erase(nums[left_idx]);
            }
            left_idx++;
        }

        total_subarrays += right_idx + 1 - left_idx;
    }
    return total_subarrays;
}

int count_k_distinct_subarrays(vector<int> &nums, int k)
{ // LeetCode Q.992.
    return count_limited_subarrays(nums, k) - count_limited_subarrays(nums, k - 1);
}
